FHEM/14_SD_UT.pm

```diff
@@ -338,1656 +340,1654 @@ use warnings;
+        } elsif ($attrValue eq 'KL_RF01') {
+          $deviceCode = substr($bitData,0,16);
+          $deviceCode = sprintf("%04X", oct( "0b$deviceCode" ) );
+          $devicename = $devicemodel.'_'.$deviceCode;
+        ############ Techmar Garden Lights ############
+        } elsif ($attrValue eq 'Techmar') {
+          $deviceCode = substr($bitData,0,32);
+          $deviceCode = sprintf("%08X", oct( "0b$deviceCode" ) );
+          $devicename = $devicemodel.'_'.$deviceCode;
+        ############ Navaris  ############
+        } elsif ($attrValue eq 'Navaris') {
+          $deviceCode = sprintf("%06X", oct( "0b$bitData" ) );
+          $devicename = $devicemodel.'_'.$deviceCode;
+        ############ unknown ############
+        } else {
+          $devicename = 'unknown_please_select_model';
+          Log3 $name, 3, "SD_UT_Attr UNDEFINED sensor $attrValue (model=unknown)";
+        }
+
+        Log3 $name, 3, "SD_UT_Attr UNDEFINED sensor $attrValue detected, code $deviceCode (DoTrigger)" if ($devicemodel ne 'unknown');
+
+        $modules{SD_UT}{defptr}{deletecache} = $name; # delete old device
+        Log3 $name, 5, "SD_UT: Attr cmd=$cmd devicename=$name attrName=$attrName attrValue=$attrValue oldmodel=$oldmodel";
+
+        readingsSingleUpdate($hash, 'state', $state, 0);
+
+        DoTrigger ('global','UNDEFINED unknown_please_select_model SD_UT unknown') if ($devicename eq 'unknown_please_select_model');   # if user push attr return to unknown
+        DoTrigger ('global',"UNDEFINED $devicename SD_UT $devicemodel $deviceCode") if ($devicename ne 'unknown_please_select_model');  # create new device
+
+        $attr{$devicename}{model} = $attrValue; # set model
+
+      } else {
+        readingsSingleUpdate($hash, 'state', 'Please press button again!', 0);
+        return "Please press button again or receive more messages!\nOnly with another message can the model be defined.\nWe need bitMSG from message.";
+      }
+    }
+  }
+
+  if ($cmd eq 'del' && $attrName eq 'model') {      ### delete readings
+    for my $readingname (qw/Button deviceCode LastAction state unknownMSG/) {
+      readingsDelete($hash,$readingname);
+    }
+  }
+
+  ## return if fhem init
+  if ($init_done) {
+    Log3 $name, 3, "SD_UT_Attr set $attrName to $attrValue" if ($cmd eq 'set');
+    return "Note: Your unknown_please_select_model device are deleted with the next receive.\nPlease use your new defined model device and do not forget to push -Save config-" if ($defs{$name}->{DEF} eq 'unknown' && $oldmodel eq 'unknown' && $attrValue ne "$oldmodel");
+  }
+  return;
 }
 
 ###################################
-sub SD_UT_bin2tristate($) {
-	my $bitData = shift;
-	my %bintotristate=(
-	 	 "00" => "0",
-	 	 "10" => "F",
-	 	 "11" => "1"
-	);
-	my $tscode;
-	for (my $n=0; $n < length($bitData); $n = $n + 2) {
-		$tscode = $tscode . $bintotristate{substr($bitData,$n,2)};
-	}
-	return $tscode;
+sub SD_UT_bin2tristate {
+  my $bitData = shift;
```

Comment:
Eingangsdaten sollen validiert werden.

z,B, 
`  my $bitData = shift; // return  $bitData`;  ?

Reply:
Die Sache mit dem shift // return ist noch Neuland.
Ich denke gegen
`my $bitData = shift; // return $bitData; `
sollte nichts sprechen. Wann tritt das return in Kraft?

Reply:
Wenn SD_UT_bin2tristate mit undef aufgerufen wird, dann hat die Variable $bitData den Wert undef.
Und im Fall von undef wird der return aufgerufen.

Die korrekte Syntax ist aber ohne Semikolon vor //  :
`my $bitData = shift // return $bitData;`

Weiterhin müsste ja verifiziert werden ob es sich um eine gerade Anzahl an Zeichen handelt oder?

Reply:
`my $bitData = shift // return $bitData;` wurde angepasst und noch weiter verifizieren so kennen wir noch nciht alle Fälle denke ich. Ich möchte ungern nun alles bestmöglich anpassen und verrammeln und dann schlägt vielleicht doch etwas mal fehl.

Reply:
Was hälst du davon ?
`  my $bitData = shift // carp 'Not called with an argument';`

`my $bitData = shift // return $bitData;`
wird dort nicht einzeln akzeptiert und wird bemängelt, das kein my $bitData  erfolgte ;-)

Da übernahm ich deinen Vorschlag und schon geht das Modul nicht mehr wegen der Warnung *smile*

Reply:
Das bringt eine Fehlermeldung, passt prinzipiell weil so transparent wird dass es hier nicht gestimmt hat:

`my $bitData = shift // carp 'Not called with an argument'; `

Schade dass es so nicht geht:
`my $bitData = shift // return $bitData;`

So dann aber bestimmt.
`my $bitData = shift // return undef;`

Reply:
Wurde angepasst.